# NAIVE BAYES CLASSIFIER

E MAIL SPAM DETECTION USING NAIVE BAYES CLASSIFIER

The Naïve Bayes classifier is a supervised machine learning algorithm, which is used for classification tasks, 
like text classification.It predicts on the basis of the probability of an object.

The algorithm is called Naïve because it assumes that the occurrence of a certain feature is 
independent of the occurrence of other features.It is based on the Bayes Theorem. 
Hence, the algorithm is called "Naïve Bayes"

STEP 1 : Import hte library that we require for this purpose

In [1]:
import pandas as pd

STEP 2: Load the dataset(in csv format) and convert it as a dataframe(df)

In [2]:
df = pd.read_csv("spam.csv")

Lets print the head of the dataframe(first 5 rows)

In [3]:
df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In the given dataset : 
    SPAM is denoted as Bad mail
    HAM is denoted as Good mail

Lets explore the data

STEP 3: Now,lets categorize the mails on spam - ham basis

In [4]:
df.groupby("Category").describe()

Message                                                            \
           count unique                                                top   
Category                                                                     
ham         4825   4516                             Sorry, I'll call later   
spam         747    641  Please call our customer service representativ...   

               
         freq  
Category       
ham        30  
spam        4

From the above result it is shown that there are 4825 ham and 747 spam mails.

Now for applying the algorithm we have to process the datas into numericals, since machines only understand numericals.

STEP 4 :For that we denote SPAM messages using '1' and HAM messages using '0'

In [5]:
df["spam"] = df["Category"].apply(lambda x: 1 if x=="spam" else 0)

STEP 5: Divide the dataframe into input(x) and output(y)

In [6]:
x=df.Message
y=df.spam

STEP 6: Lets do the train_test_split. Import the library 

In [7]:
from sklearn.model_selection import train_test_split

STEP 7: Split the data into data for training and data for testing using train_test_split method

In [8]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2)

View the trained data

In [9]:
x_train

2279    Hmm...Bad news...Hype park plaza $700 studio t...
2684    I'm okay. Chasing the dream. What's good. What...
1747    I don know account details..i will ask my mom ...
2455         Left dessert. U wan me 2 go suntec look 4 u?
4915    You've already got a flaky parent. It'snot sup...
                              ...                        
443     Oic... I saw him too but i tot he din c me... ...
5367    Just trying to figure out when I'm suppose to ...
5422              Ok k..sry i knw 2 siva..tats y i askd..
2260                 Ill call you evening ill some ideas.
2512                                  Er yep sure. Props?
Name: Message, Length: 4457, dtype: object

view the test data

In [10]:
x_test

5213                               3 pa but not selected.
984     LOOK AT THE FUCKIN TIME. WHAT THE FUCK YOU THI...
625     Nah it's straight, if you can just bring bud o...
2048    No de. But call me after some time. Ill tell y...
1046                                   Come round, it's .
                              ...                        
4682                            Are you staying in town ?
3602                         Jay told me already, will do
608     what I meant to say is cant wait to see u agai...
5202             WOT STUDENT DISCOUNT CAN U GET ON BOOKS?
4911    WE REGRET TO INFORM U THAT THE NHS HAS MADE A ...
Name: Message, Length: 1115, dtype: object

view the trained y part

In [11]:
y_train

2279    0
2684    0
1747    0
2455    0
4915    0
       ..
443     0
5367    0
5422    0
2260    0
2512    0
Name: spam, Length: 4457, dtype: int64

View the tested y part

In [12]:
y_test

5213    0
984     0
625     0
2048    0
1046    0
       ..
4682    0
3602    0
608     0
5202    0
4911    0
Name: spam, Length: 1115, dtype: int64

Still the message column has text data htat canoot be processed by computer so we have to convert it into numerical format

For that here we use count VECTORIZER method

In [13]:
from sklearn.feature_extraction.text import CountVectorizer

In [14]:
v = CountVectorizer()

STEP 8: Use fit and transform method 

In [15]:
x_train_count = v.fit_transform(x_train.values)

To convert it into an array format we use toarray() method

In [16]:
x_train_count.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

The above given is the converted array format

There are 3 types of Naive Bayes algorithm 
 1. Bernoulli's NB:if features are binary eg:gender(M/F)
 2. Multinomial NB: if features are discrete 
 3. Gausssian NB: if features are continous

here we use multinomial NB

Here, we are dealing with datas containing multiple and discrete values. Hence, we use Multinomial Naïve Bayes Algorithm.

STEP 9: Import the library 

In [17]:
from sklearn.naive_bayes import MultinomialNB

STEP 10: Create a model to check whether the algorithm is working properly

In [18]:
model = MultinomialNB()

STEP 10: train the model

In [19]:
model.fit(x_train_count,y_train)

MultinomialNB()

STEP 11:Lets do a prediction

In [20]:
email = ['i will call you later']
email_count = v.transform(email)
model.predict(email_count)

array([0], dtype=int64)

Is a ham therefore 0 will be the value

In [21]:
email = ['you won prize of 100000']
email_count = v.transform(email)
model.predict(email_count)

array([1], dtype=int64)

Is a spam therefore 1 will be the value